### PR TITLE
Update base sensor operator to support XCOM return value

### PR DESCRIPTION
**Pull request by mingshi-wang**
Inspired by the idea of https://github.com/apache/airflow/pull/20547, this PR is to update the base sensor operator such that an optional XCom value can be returned and pushed when the operator is executed. Users only need to override the poke() method to return a PokeReturnValue object that encapsulates the XCOM value.

related: https://github.com/apache/airflow/issues/20323
---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

**Review by uranusjr — APPROVED** [review]

**Comment by josh-fell** [comment]
CC @dstandish since it's in the same vein as the draft PR #20547. Maybe you've already given the thumbs-up somewhere else.

**Comment by mingshi-wang** [comment]
> CC @dstandish since it's in the same vein as the draft PR #20547. Maybe you've already given the thumbs-up somewhere else.
Hi @dstandish - can you please take a look at this PR? After this is merged, I will update https://github.com/apache/airflow/pull/20530 to be based on this PR. Thanks!

**Comment by kurtqq** [comment]
that would be nice

**Review by potiuk — APPROVED** [review]
Nice one. Also it has the potential of being detected by our "min airflow vrsion" CI check (It would fail on import).

Anyone else? @dstandish ?

File: airflow/sensors/base.py
Comment (dstandish) [inline airflow/sensors/base.py]:
```suggestion
    :param is_done: Set to true to indicate the sensor can stop poking.
    :param xcom_value: An optional XCOM value to be returned by the operator.
```

this will fix static check

**Review by dstandish — APPROVED** [review]

**Comment by dstandish** [comment]
> Nice one. Also it has the potential of being detected by our "min airflow vrsion" CI check (It would fail on import).
> 
> Anyone else? @dstandish ?

It's good with me.  I recall @potiuk that you were trying to find a way to make it possible to return a tuple.  But I think this change should be compatible with that.  What we _could_ consider doing is, for 3.0, make it so a sensor must be true / false or PRV (i.e., no longer support "truthy" values) and then in 3.1 we could add support for tuple.  WDYT?

**Comment by potiuk** [comment]
> It's good with me. I recall @potiuk that you were trying to find a way to make it possible to return a tuple. But I think this change should be compatible with that. What we _could_ consider doing is, for 3.0, make it so a sensor must be true / false or PRV (i.e., no longer support "truthy" values) and then in 3.1 we could add support for tuple. WDYT?

I am  good for it actually as is. I think we do not need to do anything more :) .
Comparing to our discussion I think the main difference is that the The PokeReturnValue with __bool__ will actually return true/false properly even for older versions of Airlfow (but it will not store xcom value).

And you should be able to start using it in our community providers even now. The way it should be done is 

```
try:
   from ... import PokeReturnValue
except ImportError:
   PokeReturnValue = None

...

    if PokeReturnValue is not None:
         return PokeReturnValue(is_done, xcom_value)
    else:
        return is_done
```

I think we will figure out the details with the first sensor which will use it (and it will be detected in CI by failing import for 2.1) and we can - then add some pre-commit checks and better information for the developers of providers (but this can be done later).

**Comment by potiuk** [comment]
Which makes me think.. Well maybe actually we should add a documentation about it. @mingshi-wang  - could you please add docs explaininig that this feature (pushing Xcom) is only available in 2.3.0  and add the example of how you can add provider compatible with both pre 2.3.0 and 2.1.0 ?

**Comment by potiuk** [comment]
Actually I realized that you CAN make it working in community providers INCLUDING XCom Push - simply the provider sensors will have to push the xcom manually until we are > 2.3.0  (which will be ~ December or so.

**Review by potiuk — CHANGES_REQUESTED** [review]
Some doc update is needed as discussed.

**Comment by mingshi-wang** [comment]
> Some doc update is needed as discussed.

Sure - I will put in some doc.

**Comment by mingshi-wang** [comment]
> 

@potiuk - I have updated the doc. Could you take a look again? This will unblock me from finishing https://github.com/apache/airflow/issues/20323. Thanks!

**Review by potiuk — APPROVED** [review]
:heart:  it

**Pull request merged**
